src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/GraphNodes/DependenciesGraphProvider.cs

```diff
@@ -271,38 +264,15 @@ internal Task TrackChangesAsync(SnapshotChangedEventArgs updatedProjectContext)
 
         private void RegisterIcons(IEnumerable<ImageMoniker> icons)
         {
-            lock (_knownIconsLock)
-            {
-                if (icons == null)
-                {
-                    return;
-                }
+            Assumes.NotNull(icons);
 
-                icons = icons.Where(x => !KnownIcons.Contains(x)).ToList();
-                if (!icons.Any())
-                {
-                    return;
-                }
-
-                KnownIcons.AddRange(icons);
-            }
-
-            ThreadHelper.JoinableTaskFactory.Run(async () =>
+            foreach (ImageMoniker icon in icons)
             {
-                IVsImageService2 imageService = null;
-                foreach (var icon in icons)
+                if (ThreadingTools.ApplyChangeOptimistically(ref _knownIcons, knownIcons => knownIcons.Add(icon)))
```

Comment:
I can't find the code for this. How does it handle races where multiple threads might be calling `Add` at the same time?

Reply:
[http://index/?leftProject=Microsoft.VisualStudio.Threading&file=ThreadingTools.cs](http://index/?leftProject=Microsoft.VisualStudio.Threading&file=ThreadingTools.cs)

---
In reply to: [157848609](https://github.com/dotnet/project-system/pull/3063#discussion_r157848609) [](ancestors = 157848609)

Reply:
Thanks - that's a handy helper to have.

Reply:
It's really quite elegant - works beautifully with immutable collections.

Reply:
This looks fine as `ImmutableHJashSet.Add` can be called from multiple threads.